// @keenthemes/ktui/lib/esm/components/range-slider/range-slider.js
/**
 * KTUI - Free & Open-Source Tailwind UI Components by Keenthemes
 * Copyright 2025 by Keenthemes Inc
 */
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
import KTData from '../../helpers/data';
import KTComponent from '../component';
var FILL_VAR = '--kt-range-fill';
var KTRangeSlider = /** @class */ (function (_super) {
    __extends(KTRangeSlider, _super);
    function KTRangeSlider(element, config) {
        if (config === void 0) { config = null; }
        var _a, _b;
        var _this = _super.call(this) || this;
        _this._name = 'range-slider';
        _this._defaultConfig = {
            output: '',
            lazy: false,
        };
        _this._config = _this._defaultConfig;
        _this._rangeInput = null;
        _this._onNativeInput = null;
        _this._onNativeChange = null;
        var input = KTRangeSlider.findRangeInput(element);
        if (!input) {
            return _this;
        }
        if (_this._shouldSkipInit(element)) {
            return _this;
        }
        _this._rangeInput = input;
        _this._init(element);
        _this._buildConfig(config);
        _this._onNativeInput = _this._handleNativeInput.bind(_this);
        _this._onNativeChange = _this._handleNativeChange.bind(_this);
        // Delegate events to the root so we keep working if the preview
        // re-renders and replaces the native range input element.
        (_a = _this._element) === null || _a === void 0 ? void 0 : _a.addEventListener('input', _this._onNativeInput);
        (_b = _this._element) === null || _b === void 0 ? void 0 : _b.addEventListener('change', _this._onNativeChange);
        _this._syncFromInput();
        return _this;
    }
    KTRangeSlider.findRangeInput = function (root) {
        if (root instanceof HTMLInputElement && root.type === 'range') {
            return root;
        }
        return root.querySelector('input[type="range"]');
    };
    KTRangeSlider.prototype._getOutputSelector = function () {
        var raw = this._getOption('output');
        return typeof raw === 'string' ? raw.trim() : '';
    };
    KTRangeSlider.prototype._resolveOutputElement = function () {
        var selector = this._getOutputSelector();
        if (!selector || !this._element) {
            return null;
        }
        var fromRoot = this._element.querySelector(selector);
        if (fromRoot) {
            return fromRoot;
        }
        return document.querySelector(selector);
    };
    KTRangeSlider.prototype._getNumericMin = function () {
        var input = this._rangeInput;
        if (!input)
            return 0;
        var n = typeof input.min === 'string' && input.min !== ''
            ? parseFloat(input.min)
            : 0;
        return Number.isFinite(n) ? n : 0;
    };
    KTRangeSlider.prototype._getNumericMax = function () {
        var input = this._rangeInput;
        if (!input)
            return 100;
        if (typeof input.max === 'string' && input.max !== '') {
            var n = parseFloat(input.max);
            if (Number.isFinite(n))
                return n;
        }
        return 100;
    };
    /** Reflects `step` for events; `undefined` when `step="any"`. */
    KTRangeSlider.prototype._getStepForPayload = function () {
        var input = this._rangeInput;
        if (!input)
            return undefined;
        var raw = input.getAttribute('step');
        if (raw === 'any')
            return undefined;
        if (raw === null || raw === '')
            return 1;
        var n = parseFloat(raw);
        return Number.isFinite(n) && n > 0 ? n : 1;
    };
    KTRangeSlider.prototype._getCurrentValue = function () {
        var input = this._rangeInput;
        if (!input)
            return 0;
        var v = typeof input.valueAsNumber === 'number' &&
            !Number.isNaN(input.valueAsNumber)
            ? input.valueAsNumber
            : parseFloat(input.value);
        return Number.isFinite(v) ? v : 0;
    };
    KTRangeSlider.prototype._clamp = function (value, min, max) {
        if (max < min) {
            return value;
        }
        return Math.min(max, Math.max(min, value));
    };
    KTRangeSlider.prototype._fillRatio = function (value, min, max) {
        if (max === min) {
            return 0;
        }
        return (this._clamp(value, min, max) - min) / (max - min);
    };
    KTRangeSlider.prototype._buildEventPayload = function () {
        var min = this._getNumericMin();
        var max = this._getNumericMax();
        var value = this._getCurrentValue();
        var step = this._getStepForPayload();
        return __assign({ value: value, min: min, max: max }, (step !== undefined ? { step: step } : {}));
    };
    KTRangeSlider.prototype._syncFromInput = function () {
        if (!this._element || !this._rangeInput) {
            return;
        }
        var min = this._getNumericMin();
        var max = this._getNumericMax();
        var value = this._getCurrentValue();
        var ratio = this._fillRatio(value, min, max);
        this._element.style.setProperty(FILL_VAR, String(ratio));
        var out = this._resolveOutputElement();
        if (out) {
            out.textContent = String(this._rangeInput.value);
        }
    };
    KTRangeSlider.prototype._handleNativeInput = function (_event) {
        var event = _event;
        var target = event.target;
        if (!(target instanceof HTMLInputElement) || target.type !== 'range') {
            return;
        }
        this._rangeInput = target;
        this._syncFromInput();
        var payload = this._buildEventPayload();
        this._fireEvent('input', payload);
        this._dispatchEvent('kt.range-slider.input', payload);
    };
    KTRangeSlider.prototype._handleNativeChange = function (_event) {
        var event = _event;
        var target = event.target;
        if (!(target instanceof HTMLInputElement) || target.type !== 'range') {
            return;
        }
        this._rangeInput = target;
        this._syncFromInput();
        var payload = this._buildEventPayload();
        this._fireEvent('change', payload);
        this._dispatchEvent('kt.range-slider.change', payload);
    };
    KTRangeSlider.prototype.getRangeInput = function () {
        return this._rangeInput;
    };
    KTRangeSlider.prototype.getValue = function () {
        return this._getCurrentValue();
    };
    KTRangeSlider.prototype.dispose = function () {
        if (this._element) {
            if (this._onNativeInput) {
                this._element.removeEventListener('input', this._onNativeInput);
            }
            if (this._onNativeChange) {
                this._element.removeEventListener('change', this._onNativeChange);
            }
        }
        this._onNativeInput = null;
        this._onNativeChange = null;
        this._rangeInput = null;
        if (this._element) {
            this._element.style.removeProperty(FILL_VAR);
        }
        _super.prototype.dispose.call(this);
    };
    KTRangeSlider.getInstance = function (element) {
        if (!element) {
            return null;
        }
        if (KTData.has(element, 'range-slider')) {
            return KTData.get(element, 'range-slider');
        }
        return null;
    };
    KTRangeSlider.getOrCreateInstance = function (element, config) {
        var existing = this.getInstance(element);
        if (existing) {
            return existing;
        }
        if (!this.findRangeInput(element)) {
            return null;
        }
        new KTRangeSlider(element, config !== null && config !== void 0 ? config : undefined);
        return this.getInstance(element);
    };
    KTRangeSlider.createInstances = function () {
        document
            .querySelectorAll('[data-kt-range-slider]')
            .forEach(function (el) {
            if (el.getAttribute('data-kt-range-slider-lazy') === 'true') {
                return;
            }
            new KTRangeSlider(el);
        });
    };
    KTRangeSlider.init = function () {
        KTRangeSlider.createInstances();
    };
    return KTRangeSlider;
}(KTComponent));
export { KTRangeSlider };
if (typeof window !== 'undefined') {
    window.KTRangeSlider = KTRangeSlider;
}
