// @keenthemes/ktui/lib/cjs/components/repeater/repeater.js
"use strict";
/**
 * KTUI - Free & Open-Source Tailwind UI Components by Keenthemes
 * Copyright 2025 by Keenthemes Inc
 */
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.KTRepeater = void 0;
var data_1 = require("../../helpers/data");
var dom_1 = require("../../helpers/dom");
var component_1 = require("../component");
var DELETE_SELECTOR = '[data-kt-repeater-delete]';
var KTRepeater = /** @class */ (function (_super) {
    __extends(KTRepeater, _super);
    function KTRepeater(element, config) {
        if (config === void 0) { config = null; }
        var _this = _super.call(this) || this;
        _this._name = 'repeater';
        _this._defaultConfig = {
            target: '',
            wrapper: '',
            limit: 0,
        };
        _this._config = _this
            ._defaultConfig;
        _this._wrapperElement = null;
        if (data_1.default.has(element, _this._name))
            return _this;
        _this._init(element);
        _this._buildConfig(config);
        if (!_this._element)
            return _this;
        _this._wrapperElement = dom_1.default.getElement(_this._getOption('wrapper'));
        if (!_this._wrapperElement)
            return _this;
        _this._deleteHandler = _this._handleDeleteClick.bind(_this);
        _this._handlers();
        _this._syncTriggerDisabled();
        _this._syncDeleteButtonsDisabled();
        return _this;
    }
    KTRepeater.prototype._getTargetElement = function () {
        return dom_1.default.getElement(this._getOption('target'));
    };
    KTRepeater.prototype._getCloneCount = function () {
        return this._wrapperElement ? this._wrapperElement.children.length : 0;
    };
    KTRepeater.prototype._syncTriggerDisabled = function () {
        if (!this._element || !this._wrapperElement)
            return;
        var limit = Number(this._getOption('limit')) || 0;
        if (limit <= 0)
            return;
        var atLimit = this._getCloneCount() >= limit;
        if (this._element.hasAttribute('disabled') !== atLimit) {
            if (atLimit) {
                this._element.setAttribute('disabled', '');
            }
            else {
                this._element.removeAttribute('disabled');
            }
        }
    };
    KTRepeater.prototype._syncDeleteButtonsDisabled = function () {
        if (!this._wrapperElement)
            return;
        var count = this._wrapperElement.children.length;
        var isOnlyOne = count <= 1;
        Array.from(this._wrapperElement.children).forEach(function (row) {
            var deleteBtn = row.querySelector(DELETE_SELECTOR);
            if (!deleteBtn)
                return;
            if (isOnlyOne) {
                deleteBtn.setAttribute('disabled', '');
                deleteBtn.setAttribute('aria-disabled', 'true');
            }
            else {
                deleteBtn.removeAttribute('disabled');
                deleteBtn.removeAttribute('aria-disabled');
            }
        });
    };
    KTRepeater.prototype._handlers = function () {
        var _this = this;
        var _a;
        if (!this._element)
            return;
        this._element.addEventListener('click', function (event) {
            event.preventDefault();
            _this._add();
        });
        (_a = this._wrapperElement) === null || _a === void 0 ? void 0 : _a.addEventListener('click', this._deleteHandler);
    };
    KTRepeater.prototype._handleDeleteClick = function (event) {
        var target = event.target;
        var deleteEl = target.closest(DELETE_SELECTOR);
        if (!deleteEl || !this._wrapperElement)
            return;
        // Keep at least one row so the template remains for cloning
        if (this._wrapperElement.children.length <= 1)
            return;
        var cloneContainer = Array.from(this._wrapperElement.children).find(function (child) { return child.contains(deleteEl); });
        if (!cloneContainer)
            return;
        dom_1.default.remove(cloneContainer);
        this._syncTriggerDisabled();
        this._syncDeleteButtonsDisabled();
    };
    KTRepeater.prototype._add = function () {
        if (!this._wrapperElement)
            return;
        var targetEl = this._getTargetElement();
        if (!targetEl)
            return;
        var limit = Number(this._getOption('limit')) || 0;
        var count = this._getCloneCount();
        if (limit > 0 && count >= limit)
            return;
        var clone = targetEl.cloneNode(true);
        this._wrapperElement.appendChild(clone);
        var payload = { element: clone };
        this._fireEvent('add', payload);
        this._dispatchEvent('add', payload);
        this._syncTriggerDisabled();
        this._syncDeleteButtonsDisabled();
    };
    KTRepeater.prototype.add = function () {
        this._add();
    };
    KTRepeater.getInstance = function (element) {
        if (!element)
            return null;
        if (data_1.default.has(element, 'repeater')) {
            return data_1.default.get(element, 'repeater');
        }
        return null;
    };
    KTRepeater.getOrCreateInstance = function (element, config) {
        return (this.getInstance(element) || new KTRepeater(element, config !== null && config !== void 0 ? config : undefined));
    };
    KTRepeater.createInstances = function () {
        document.querySelectorAll('[data-kt-repeater]').forEach(function (el) {
            new KTRepeater(el);
        });
    };
    KTRepeater.init = function () {
        KTRepeater.createInstances();
    };
    KTRepeater.prototype.dispose = function () {
        var _a;
        (_a = this._wrapperElement) === null || _a === void 0 ? void 0 : _a.removeEventListener('click', this._deleteHandler);
        _super.prototype.dispose.call(this);
    };
    return KTRepeater;
}(component_1.default));
exports.KTRepeater = KTRepeater;
if (typeof window !== 'undefined') {
    window.KTRepeater = KTRepeater;
}
