// @keenthemes/ktui/lib/cjs/components/drawer/drawer.js
"use strict";
/**
 * KTUI - Free & Open-Source Tailwind UI Components by Keenthemes
 * Copyright 2025 by Keenthemes Inc
 */
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __spreadArray = (this && this.__spreadArray) || function (to, from, pack) {
    if (pack || arguments.length === 2) for (var i = 0, l = from.length, ar; i < l; i++) {
        if (ar || !(i in from)) {
            if (!ar) ar = Array.prototype.slice.call(from, 0, i);
            ar[i] = from[i];
        }
    }
    return to.concat(ar || Array.prototype.slice.call(from));
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.KTDrawer = void 0;
var data_1 = require("../../helpers/data");
var dom_1 = require("../../helpers/dom");
var utils_1 = require("../../helpers/utils");
var event_handler_1 = require("../../helpers/event-handler");
var component_1 = require("../component");
var KTDrawer = /** @class */ (function (_super) {
    __extends(KTDrawer, _super);
    function KTDrawer(element, config) {
        var _this = _super.call(this) || this;
        _this._name = 'drawer';
        _this._defaultConfig = {
            zindex: '100',
            enable: true,
            class: '',
            shownClass: 'flex',
            hiddenClass: 'hidden',
            backdrop: true,
            backdropClass: 'kt-drawer-backdrop',
            backdropStatic: false,
            keyboard: true,
            disableScroll: true,
            persistent: false,
            container: '',
            focus: true,
            keepInPlaceWithin: '',
        };
        _this._config = _this._defaultConfig;
        _this._isOpen = false;
        _this._isTransitioning = false;
        _this._backdropElement = null;
        _this._relatedTarget = null;
        if (data_1.default.has(element, _this._name)) {
            return _this;
        }
        _this._init(element);
        _this._buildConfig(config);
        _this._handleClose();
        _this._update();
        _this._handleContainer();
        return _this;
    }
    KTDrawer.prototype._handleClose = function () {
        var _this = this;
        if (!this._element)
            return;
        event_handler_1.default.on(this._element, '[data-kt-drawer-hide]', 'click', function () {
            _this._hide();
        });
    };
    KTDrawer.prototype._toggle = function (relatedTarget) {
        var payload = { cancel: false };
        this._fireEvent('toggle', payload);
        this._dispatchEvent('toggle', payload);
        if (payload.cancel === true) {
            return;
        }
        if (this._isOpen === true) {
            this._hide();
        }
        else {
            this._show(relatedTarget);
        }
    };
    KTDrawer.prototype._show = function (relatedTarget) {
        var _this = this;
        if (this._isOpen || this._isTransitioning) {
            return;
        }
        var payload = { cancel: false };
        this._fireEvent('show', payload);
        this._dispatchEvent('show', payload);
        if (payload.cancel === true) {
            return;
        }
        KTDrawer.hide();
        // When container="body", move drawer to body only if NOT inside an element matching keepInPlaceWithin.
        // When keepInPlaceWithin is set (e.g. for SPA/persisted layouts), keeping the drawer in place lets the host preserve it across navigations.
        if (this._getOption('container') === 'body' &&
            this._element.parentElement !== document.body) {
            var keepInPlace = this._isKeepInPlace();
            if (!keepInPlace) {
                if (!this._element.hasAttribute('data-kt-drawer-original-parent-id')) {
                    var originalParent = this._element.parentElement;
                    if (originalParent && originalParent !== document.body) {
                        this._element.setAttribute('data-kt-drawer-original-parent-id', originalParent.id || '');
                    }
                }
                document.body.appendChild(this._element);
            }
        }
        if (this._getOption('backdrop') === true)
            this._createBackdrop();
        if (relatedTarget)
            this._relatedTarget = relatedTarget;
        if (!this._element)
            return;
        this._isTransitioning = true;
        this._element.classList.remove(this._getOption('hiddenClass'));
        this._element.classList.add(this._getOption('shownClass'));
        this._element.setAttribute('role', 'dialog');
        this._element.setAttribute('aria-modal', 'true');
        this._element.setAttribute('tabindex', '-1');
        var zindex = parseInt(this._getOption('zindex'));
        if (zindex > 0) {
            this._element.style.zIndex = "".concat(zindex);
        }
        if (this._getOption('disableScroll')) {
            document.body.style.overflow = 'hidden';
        }
        dom_1.default.reflow(this._element);
        this._element.classList.add('open');
        dom_1.default.transitionEnd(this._element, function () {
            _this._isTransitioning = false;
            _this._isOpen = true;
            if (_this._getOption('focus') === true) {
                _this._autoFocus();
            }
            _this._fireEvent('shown');
            _this._dispatchEvent('shown');
        });
    };
    KTDrawer.prototype._hide = function () {
        var _this = this;
        if (!this._element)
            return;
        if (this._isOpen === false || this._isTransitioning) {
            return;
        }
        var payload = { cancel: false };
        this._fireEvent('hide', payload);
        this._dispatchEvent('hide', payload);
        if (payload.cancel === true) {
            return;
        }
        this._isTransitioning = true;
        this._element.removeAttribute('role');
        this._element.removeAttribute('aria-modal');
        this._element.removeAttribute('tabindex');
        if (this._getOption('disableScroll')) {
            document.body.style.overflow = '';
        }
        dom_1.default.reflow(this._element);
        this._element.classList.remove('open');
        if (this._getOption('backdrop') === true) {
            this._deleteBackdrop();
        }
        dom_1.default.transitionEnd(this._element, function () {
            if (!_this._element)
                return;
            _this._isTransitioning = false;
            _this._isOpen = false;
            _this._element.classList.add(_this._getOption('hiddenClass'));
            _this._element.classList.remove(_this._getOption('shownClass'));
            _this._element.style.zIndex = '';
            // Note: We don't move drawer back to original location here
            // Livewire will handle DOM structure on next navigation, and drawer will be reinitialized
            // in its original location from the persisted component HTML
            _this._fireEvent('hidden');
            _this._dispatchEvent('hidden');
        });
    };
    KTDrawer.prototype._update = function () {
        var _a;
        if (((_a = this._getOption('class')) === null || _a === void 0 ? void 0 : _a.length) > 0) {
            if (this.isEnabled()) {
                dom_1.default.addClass(this._element, this._getOption('class'));
            }
            else {
                dom_1.default.removeClass(this._element, this._getOption('class'));
            }
        }
    };
    KTDrawer.prototype._handleContainer = function () {
        var _a;
        if (this._getOption('container')) {
            if (this._getOption('container') === 'body') {
                if (this._isKeepInPlace()) {
                    if (!this._element.style.position ||
                        this._element.style.position === 'static') {
                        this._element.style.position = 'fixed';
                    }
                }
                else {
                    document.body.appendChild(this._element);
                }
            }
            else {
                (_a = document
                    .querySelector(this._getOption('container'))) === null || _a === void 0 ? void 0 : _a.appendChild(this._element);
            }
        }
    };
    /** True when drawer is inside an element matching keepInPlaceWithin (so we keep it in place instead of moving to body). */
    KTDrawer.prototype._isKeepInPlace = function () {
        var _a, _b;
        var selector = (_a = this._getOption('keepInPlaceWithin')) === null || _a === void 0 ? void 0 : _a.trim();
        if (!selector || !((_b = this._element) === null || _b === void 0 ? void 0 : _b.parentElement))
            return false;
        var parent = this._element.parentElement;
        var selectors = selector
            .split(',')
            .map(function (s) { return s.trim(); })
            .filter(Boolean);
        for (var _i = 0, selectors_1 = selectors; _i < selectors_1.length; _i++) {
            var sel = selectors_1[_i];
            try {
                if (parent.closest(sel) !== null)
                    return true;
            }
            catch (_c) {
                // invalid selector, skip
            }
        }
        return false;
    };
    KTDrawer.prototype._autoFocus = function () {
        if (!this._element)
            return;
        var input = this._element.querySelector('[data-kt-drawer-focus]');
        if (!input)
            return;
        else
            input.focus();
    };
    KTDrawer.prototype._createBackdrop = function () {
        var _this = this;
        if (!this._element)
            return;
        var zindex = parseInt(this._getOption('zindex'));
        this._backdropElement = document.createElement('DIV');
        this._backdropElement.style.zIndex = (zindex - 1).toString();
        this._backdropElement.setAttribute('data-kt-drawer-backdrop', 'true');
        var parent = this._element.parentElement;
        if (parent) {
            parent.insertBefore(this._backdropElement, this._element);
        }
        else {
            document.body.append(this._backdropElement);
        }
        dom_1.default.reflow(this._backdropElement);
        dom_1.default.addClass(this._backdropElement, this._getOption('backdropClass'));
        this._backdropElement.addEventListener('click', function (event) {
            event.preventDefault();
            if (_this._getOption('backdropStatic') === false) {
                _this._hide();
            }
        });
    };
    KTDrawer.prototype._deleteBackdrop = function () {
        var _this = this;
        if (!this._backdropElement)
            return;
        dom_1.default.reflow(this._backdropElement);
        this._backdropElement.style.opacity = '0';
        dom_1.default.transitionEnd(this._backdropElement, function () {
            if (!_this._backdropElement)
                return;
            dom_1.default.remove(_this._backdropElement);
        });
    };
    KTDrawer.prototype._isEnabled = function () {
        return utils_1.default.stringToBoolean(this._getOption('enable'));
    };
    KTDrawer.prototype.toggle = function (relatedTarget) {
        return this._toggle(relatedTarget);
    };
    KTDrawer.prototype.show = function (relatedTarget) {
        return this._show(relatedTarget);
    };
    KTDrawer.prototype.hide = function () {
        return this._hide();
    };
    KTDrawer.prototype.update = function () {
        return this._update();
    };
    KTDrawer.prototype.getRelatedTarget = function () {
        return this._relatedTarget;
    };
    KTDrawer.prototype.isOpen = function () {
        return this._isOpen;
    };
    KTDrawer.prototype.isEnabled = function () {
        return this._isEnabled();
    };
    KTDrawer.getElement = function (reference) {
        if (reference && reference.hasAttribute('data-kt-drawer-initialized'))
            return reference;
        var findElement = reference &&
            reference.closest('[data-kt-drawer-initialized]');
        if (findElement)
            return findElement;
        // Fallback: look for parent with data-kt-drawer attribute
        if (reference) {
            var drawerContainer = reference.closest('[data-kt-drawer]');
            if (drawerContainer)
                return drawerContainer;
        }
        // If reference is a toggle button with a selector, find the drawer by selector
        // This handles cases where the toggle button is not a child of the drawer
        if (reference && reference.hasAttribute('data-kt-drawer-toggle')) {
            var selector = reference.getAttribute('data-kt-drawer-toggle');
            if (selector) {
                // Check both document and body (drawers with container="body" are moved to body)
                var drawerElInDoc = document.querySelector(selector);
                var drawerElInBody = document.body.querySelector(selector);
                var drawerEl = drawerElInDoc || drawerElInBody;
                if (drawerEl)
                    return drawerEl;
            }
        }
        return null;
    };
    /**
     * Wait for an element to appear in the DOM using polling with MutationObserver fallback
     * Useful for persisted Livewire components that may not be in DOM immediately
     */
    KTDrawer.waitForElement = function (selector, timeout) {
        if (timeout === void 0) { timeout = 2000; }
        return new Promise(function (resolve) {
            var resolved = false;
            var doResolve = function (element) {
                if (!resolved) {
                    resolved = true;
                    resolve(element);
                }
            };
            // Check if element already exists
            var existing = document.querySelector(selector) ||
                document.body.querySelector(selector);
            if (existing) {
                doResolve(existing);
                return;
            }
            // Use polling for faster detection (check every 50ms)
            var attempts = 0;
            var maxAttempts = timeout / 50;
            var pollInterval = setInterval(function () {
                if (resolved) {
                    clearInterval(pollInterval);
                    return;
                }
                attempts++;
                var element = document.querySelector(selector) ||
                    document.body.querySelector(selector);
                if (element) {
                    clearInterval(pollInterval);
                    doResolve(element);
                    return;
                }
                if (attempts >= maxAttempts) {
                    clearInterval(pollInterval);
                    doResolve(null);
                }
            }, 50);
            // Also use MutationObserver as backup for immediate detection
            var observer = new MutationObserver(function () {
                if (resolved) {
                    observer.disconnect();
                    return;
                }
                var element = document.querySelector(selector) ||
                    document.body.querySelector(selector);
                if (element) {
                    clearInterval(pollInterval);
                    observer.disconnect();
                    doResolve(element);
                }
            });
            observer.observe(document.body, {
                childList: true,
                subtree: true,
            });
            // Cleanup on timeout
            setTimeout(function () {
                if (!resolved) {
                    clearInterval(pollInterval);
                    observer.disconnect();
                    doResolve(null);
                }
            }, timeout);
        });
    };
    KTDrawer.getInstance = function (element) {
        if (!element)
            return null;
        var drawerElement = KTDrawer.getElement(element);
        if (!drawerElement) {
            // If element is a toggle button and drawer element wasn't found, return null
            // The handleToggle() will handle waiting for the element to appear
            if (element.hasAttribute('data-kt-drawer-toggle')) {
            }
            return null;
        }
        if (data_1.default.has(drawerElement, 'drawer')) {
            return data_1.default.get(drawerElement, 'drawer');
        }
        if (drawerElement.getAttribute('data-kt-drawer-initialized') === 'true') {
            return new KTDrawer(drawerElement);
        }
        return null;
    };
    KTDrawer.getOrCreateInstance = function (element, config) {
        return this.getInstance(element) || new KTDrawer(element, config);
    };
    KTDrawer.hide = function () {
        var elements = document.querySelectorAll('[data-kt-drawer-initialized]');
        elements.forEach(function (element) {
            var drawer = KTDrawer.getInstance(element);
            if (drawer && drawer.isOpen()) {
                drawer.hide();
            }
        });
    };
    KTDrawer.handleResize = function () {
        window.addEventListener('resize', function () {
            var timer;
            utils_1.default.throttle(timer, function () {
                document
                    .querySelectorAll('[data-kt-drawer-initialized]')
                    .forEach(function (element) {
                    var drawer = KTDrawer.getInstance(element);
                    drawer.update();
                    if (drawer && drawer.isOpen() && !drawer.isEnabled()) {
                        drawer.hide();
                    }
                });
            }, 200);
        });
    };
    KTDrawer.handleToggle = function () {
        // Add raw click listener to document.body to track all clicks
        document.body.addEventListener('click', function (rawEvent) {
            var target = rawEvent.target;
            if (target && target.hasAttribute('data-kt-drawer-toggle')) {
            }
        }, true); // Use capture phase to catch before any stopPropagation
        event_handler_1.default.on(document.body, '[data-kt-drawer-toggle]', 'click', function (event, target) {
            var _a;
            event.stopPropagation();
            var selector = target.getAttribute('data-kt-drawer-toggle');
            if (!selector)
                return;
            // Try to get instance immediately
            var drawer = KTDrawer.getInstance(target);
            if (drawer) {
                drawer.toggle(target);
            }
            else {
                // Drawer element not found - wait for it to appear (handles persisted Livewire components)
                // Check if drawer exists in persisted components (might be in header that's persisted)
                var persistedHeader = ((_a = document.querySelector('[wire\\:id]')) === null || _a === void 0 ? void 0 : _a.closest('[wire\\:id]')) ||
                    document.querySelector('header#header');
                var drawerInPersisted = persistedHeader
                    ? persistedHeader.querySelector(selector)
                    : null;
                // Wait longer for persisted components that may take time to render
                // Also check if drawer exists in persisted header component
                KTDrawer.waitForElement(selector, 5000).then(function (drawerElement) {
                    if (drawerElement) {
                        // Initialize the drawer if not already initialized
                        if (!data_1.default.has(drawerElement, 'drawer')) {
                            new KTDrawer(drawerElement);
                        }
                        // Get instance and toggle
                        var drawerInstance = KTDrawer.getInstance(drawerElement);
                        if (drawerInstance) {
                            drawerInstance.toggle(target);
                        }
                    }
                    else {
                        // Drawer never appeared - trigger a reinit to see if it helps
                        // This handles cases where drawers are in persisted components that haven't rendered yet
                        setTimeout(function () {
                            KTDrawer.reinit();
                            // Try one more time after reinit
                            var drawerAfterReinit = document.querySelector(selector) ||
                                document.body.querySelector(selector);
                            if (drawerAfterReinit) {
                                if (!data_1.default.has(drawerAfterReinit, 'drawer')) {
                                    new KTDrawer(drawerAfterReinit);
                                }
                                var drawerInstance = KTDrawer.getInstance(drawerAfterReinit);
                                if (drawerInstance) {
                                    drawerInstance.toggle(target);
                                }
                            }
                        }, 500);
                    }
                });
            }
        });
    };
    KTDrawer.handleDismiss = function () {
        event_handler_1.default.on(document.body, '[data-kt-drawer-dismiss]', 'click', function (event, target) {
            event.stopPropagation();
            var modalElement = target.closest('[data-kt-drawer="true"]');
            if (modalElement) {
                var modal = KTDrawer.getInstance(modalElement);
                if (modal) {
                    modal.hide();
                }
            }
        });
    };
    KTDrawer.handleClickAway = function () {
        document.addEventListener('click', function (event) {
            var drawerEl = document.querySelector('.open[data-kt-drawer-initialized]');
            if (!drawerEl)
                return;
            var drawer = KTDrawer.getInstance(drawerEl);
            if (!drawer)
                return;
            if (drawer.getOption('persistent'))
                return;
            if (drawer.getOption('backdrop'))
                return;
            if (drawerEl !== event.target &&
                drawer.getRelatedTarget() !== event.target &&
                drawerEl.contains(event.target) === false) {
                drawer.hide();
            }
        });
    };
    KTDrawer.handleKeyword = function () {
        document.addEventListener('keydown', function (event) {
            var drawerEl = document.querySelector('.open[data-kt-drawer-initialized]');
            var drawer = KTDrawer.getInstance(drawerEl);
            if (!drawer) {
                return;
            }
            // if esc key was not pressed in combination with ctrl or alt or shift
            if (event.key === 'Escape' &&
                !(event.ctrlKey || event.altKey || event.shiftKey)) {
                drawer.hide();
            }
            if (event.code === 'Tab' && !event.metaKey) {
                return;
            }
        });
    };
    KTDrawer.createInstances = function () {
        // Find all drawer elements - check both document and body (drawers with container="body" are moved there)
        var elementsInDoc = document.querySelectorAll('[data-kt-drawer]');
        var elementsInBody = document.body.querySelectorAll('[data-kt-drawer]');
        // Combine and deduplicate
        var allElements = new Set(__spreadArray(__spreadArray([], Array.from(elementsInDoc), true), Array.from(elementsInBody), true));
        var elements = Array.from(allElements);
        elements.forEach(function (element) {
            new KTDrawer(element);
        });
    };
    KTDrawer.init = function () {
        KTDrawer.createInstances();
        if (window.KT_DRAWER_INITIALIZED !== true) {
            KTDrawer.handleToggle();
            KTDrawer.handleDismiss();
            KTDrawer.handleResize();
            KTDrawer.handleClickAway();
            KTDrawer.handleKeyword();
            window.KT_DRAWER_INITIALIZED = true;
        }
    };
    /**
     * Force reinitialization of drawers by clearing KTData entries.
     * Useful for Livewire wire:navigate where persisted elements need reinitialization.
     */
    KTDrawer.reinit = function () {
        // Follow the same simple pattern as KTDropdown.reinit()
        // Find all drawer elements - check both document and body (some may be moved to body)
        var elementsInDoc = document.querySelectorAll('[data-kt-drawer]');
        var elementsInBody = document.body.querySelectorAll('[data-kt-drawer]');
        // Combine and deduplicate
        var allElements = new Set(__spreadArray(__spreadArray([], Array.from(elementsInDoc), true), Array.from(elementsInBody), true));
        var elements = Array.from(allElements);
        // Clean up existing instances
        elements.forEach(function (element) {
            try {
                // Get existing instance to clean up
                var instance = KTDrawer.getInstance(element);
                if (instance && typeof instance.hide === 'function') {
                    instance.hide(); // This will clean up backdrop and state
                }
                // Clear KTData entries
                var hadDrawer = data_1.default.has(element, 'drawer');
                data_1.default.remove(element, 'drawer');
                // Remove initialization attribute to allow fresh initialization
                element.removeAttribute('data-kt-drawer-initialized');
            }
            catch (e) {
                // Ignore errors for individual elements
            }
        });
        // Now create fresh instances
        KTDrawer.createInstances();
        // Always ensure handlers are set up (similar to KTMenu.init() behavior)
        // Event handlers use delegation so they persist, but we ensure they're attached
        KTDrawer.handleToggle();
        KTDrawer.handleDismiss();
        KTDrawer.handleResize();
        KTDrawer.handleClickAway();
        KTDrawer.handleKeyword();
    };
    return KTDrawer;
}(component_1.default));
exports.KTDrawer = KTDrawer;
if (typeof window !== 'undefined') {
    window.KTDrawer = KTDrawer;
}
