// @keenthemes/ktui/lib/esm/components/rating/rating.js
/**
 * KTUI - Free & Open-Source Tailwind UI Components by Keenthemes
 * Copyright 2025 by Keenthemes Inc
 */
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __spreadArray = (this && this.__spreadArray) || function (to, from, pack) {
    if (pack || arguments.length === 2) for (var i = 0, l = from.length, ar; i < l; i++) {
        if (ar || !(i in from)) {
            if (!ar) ar = Array.prototype.slice.call(from, 0, i);
            ar[i] = from[i];
        }
    }
    return to.concat(ar || Array.prototype.slice.call(from));
};
import KTData from '../../helpers/data';
import KTComponent from '../component';
var STAR_SVG = '<path d="M3.612 15.443c-.386.198-.824-.149-.746-.592l.83-4.73L.173 6.765c-.329-.314-.158-.888.283-.95l4.898-.696L7.538.792c.197-.39.73-.39.927 0l2.184 4.327 4.898.696c.441.062.612.636.282.95l-3.522 3.356.83 4.73c.078.443-.36.79-.746.592L8 13.187l-4.389 2.256z"/>';
var HEART_SVG = '<path fill-rule="evenodd" d="M8 1.314C12.438-3.248 23.534 4.735 8 15-7.534 4.736 3.562-3.248 8 1.314z"/>';
var KTRating = /** @class */ (function (_super) {
    __extends(KTRating, _super);
    function KTRating(element, config) {
        if (config === void 0) { config = null; }
        var _this = _super.call(this) || this;
        _this._name = 'rating';
        _this._defaultConfig = {
            value: 0,
            max: 5,
            readonly: false,
            name: 'rating',
            symbol: 'star',
            lazy: false,
        };
        _this._config = _this._defaultConfig;
        _this._container = null;
        _this._changeListener = null;
        if (KTData.has(element, _this._name))
            return _this;
        _this._init(element);
        _this._buildConfig(config);
        _this._render();
        if (!_this._config.readonly) {
            _this._handlers();
        }
        return _this;
    }
    KTRating.prototype._getMax = function () {
        var max = Number(this._getOption('max'));
        return Number.isFinite(max) && max >= 1 ? Math.floor(max) : 5;
    };
    KTRating.prototype._getValue = function () {
        var v = Number(this._getOption('value'));
        var max = this._getMax();
        return Number.isFinite(v) && v >= 0 && v <= max ? Math.floor(v) : 0;
    };
    KTRating.prototype._render = function () {
        if (!this._element)
            return;
        var max = this._getMax();
        var value = this._getValue();
        var readonly = this._config.readonly === true;
        var symbol = this._config.symbol || 'star';
        var name = this._config.name || 'rating';
        var isStar = symbol !== 'heart';
        var filledClass = isStar
            ? 'text-yellow-400 dark:text-yellow-600'
            : 'text-red-500 dark:text-red-500';
        var unfilledClass = 'text-muted-foreground/50 dark:text-muted-foreground/50';
        var path = isStar ? STAR_SVG : HEART_SVG;
        var svg = "<svg class=\"shrink-0 size-5\" xmlns=\"http://www.w3.org/2000/svg\" width=\"16\" height=\"16\" fill=\"currentColor\" viewBox=\"0 0 16 16\">".concat(path, "</svg>");
        var container = document.createElement('div');
        container.className =
            'kt-rating flex flex-row-reverse justify-end items-center gap-0';
        container.setAttribute('role', readonly ? 'img' : 'group');
        container.setAttribute('aria-label', "Rating: ".concat(value, " of ").concat(max));
        if (readonly) {
            container.setAttribute('aria-valuenow', String(value));
            container.setAttribute('aria-valuemin', '0');
            container.setAttribute('aria-valuemax', String(max));
        }
        if (readonly) {
            for (var i = max; i >= 1; i--) {
                var filled = i <= value;
                var span = document.createElement('span');
                span.className = filled ? filledClass : unfilledClass;
                span.innerHTML = svg;
                container.appendChild(span);
            }
        }
        else {
            container.setAttribute('aria-valuenow', String(value));
            container.setAttribute('aria-valuemin', '1');
            container.setAttribute('aria-valuemax', String(max));
            for (var i = max; i >= 1; i--) {
                var id = "kt-rating-".concat(this._uid, "-").concat(i);
                var radio = document.createElement('input');
                radio.type = 'radio';
                radio.name = name;
                radio.value = String(i);
                radio.id = id;
                radio.className =
                    'peer -ms-5 size-5 bg-transparent border-0 text-transparent cursor-pointer appearance-none checked:bg-none focus:bg-none focus:ring-0 focus:ring-offset-0';
                radio.setAttribute('aria-label', "Rate ".concat(i, " of ").concat(max));
                if (i === value)
                    radio.checked = true;
                var label = document.createElement('label');
                label.htmlFor = id;
                label.setAttribute('data-kt-rating-value', String(i));
                label.className = "cursor-pointer kt-rating-label ".concat(i <= value ? filledClass : unfilledClass);
                label.innerHTML = svg;
                container.appendChild(radio);
                container.appendChild(label);
            }
            this._updateInteractiveDisplay();
        }
        this._element.innerHTML = '';
        this._element.appendChild(container);
        this._container = container;
    };
    KTRating.prototype._updateInteractiveDisplay = function () {
        if (!this._container || this._config.readonly)
            return;
        var val = this.getValue();
        var max = this._getMax();
        var isStar = this._config.symbol !== 'heart';
        var filledClass = isStar
            ? 'text-yellow-400 dark:text-yellow-600'
            : 'text-red-500 dark:text-red-500';
        var unfilledClass = 'text-muted-foreground/50 dark:text-muted-foreground/50';
        var filledTokens = filledClass.split(' ');
        var unfilledTokens = unfilledClass.split(' ');
        this._container
            .querySelectorAll('.kt-rating-label')
            .forEach(function (label) {
            var _a, _b;
            var v = parseInt(label.getAttribute('data-kt-rating-value') || '0', 10);
            var filled = v <= (val !== null && val !== void 0 ? val : 0);
            (_a = label.classList).remove.apply(_a, __spreadArray(__spreadArray([], filledTokens, false), unfilledTokens, false));
            (_b = label.classList).add.apply(_b, (filled ? filledTokens : unfilledTokens));
        });
        this._container.setAttribute('aria-valuenow', val != null ? String(val) : '0');
        this._container.setAttribute('aria-label', "Rating: ".concat(val !== null && val !== void 0 ? val : 0, " of ").concat(max));
    };
    KTRating.prototype._handlers = function () {
        var _this = this;
        if (!this._container)
            return;
        this._changeListener = function () {
            _this._updateInteractiveDisplay();
            var val = _this.getValue();
            _this._fireEvent('change', { value: val });
            _this._dispatchEvent('kt.rating.change', { value: val });
        };
        this._container.addEventListener('change', this._changeListener);
    };
    KTRating.prototype.getValue = function () {
        var _a;
        if (!this._element)
            return null;
        if (this._config.readonly) {
            var v = this._getValue();
            return v > 0 ? v : null;
        }
        var radio = (_a = this._container) === null || _a === void 0 ? void 0 : _a.querySelector('input[type="radio"]:checked');
        if (!radio)
            return null;
        var n = parseInt(radio.value, 10);
        return Number.isFinite(n) ? n : null;
    };
    KTRating.prototype.setValue = function (value) {
        if (!this._container)
            return;
        var max = this._getMax();
        if (this._config.readonly)
            return;
        if (value !== null && value >= 1 && value <= max) {
            var radio = this._container.querySelector("input[type=\"radio\"][value=\"".concat(value, "\"]"));
            if (radio) {
                radio.checked = true;
                this._updateInteractiveDisplay();
            }
        }
        else {
            this._container
                .querySelectorAll('input[type="radio"]')
                .forEach(function (r) {
                r.checked = false;
            });
            this._updateInteractiveDisplay();
        }
    };
    KTRating.prototype.dispose = function () {
        if (this._container && this._changeListener) {
            this._container.removeEventListener('change', this._changeListener);
            this._changeListener = null;
        }
        this._container = null;
        _super.prototype.dispose.call(this);
    };
    KTRating.getInstance = function (element) {
        if (!element)
            return null;
        if (KTData.has(element, 'rating')) {
            return KTData.get(element, 'rating');
        }
        if (element.getAttribute('data-kt-rating') !== null) {
            return new KTRating(element);
        }
        return null;
    };
    KTRating.createInstances = function () {
        var elements = document.querySelectorAll('[data-kt-rating]');
        elements.forEach(function (el) {
            if (el.getAttribute('data-kt-rating-lazy') === 'true')
                return;
            new KTRating(el);
        });
    };
    KTRating.init = function () {
        KTRating.createInstances();
    };
    return KTRating;
}(KTComponent));
export { KTRating };
if (typeof window !== 'undefined') {
    window.KTRating = KTRating;
}
