// @keenthemes/ktui/lib/cjs/components/datatable/datatable-sort.js
"use strict";
/**
 * KTUI - Free & Open-Source Tailwind UI Components by Keenthemes
 * Copyright 2025 by Keenthemes Inc
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.createSortHandler = createSortHandler;
function createSortHandler(config, theadElement, getState, setState, fireEvent, dispatchEvent, updateData) {
    // Helper to compare values for sorting (string)
    function compareValues(a, b, sortOrder) {
        var aText = String(a).replace(/<[^>]*>|&nbsp;/g, '');
        var bText = String(b).replace(/<[^>]*>|&nbsp;/g, '');
        return aText > bText
            ? sortOrder === 'asc'
                ? 1
                : -1
            : aText < bText
                ? sortOrder === 'asc'
                    ? -1
                    : 1
                : 0;
    }
    // Parse value for numeric sort: strip currency/commas, then parseFloat
    function parseNumeric(value) {
        if (value === null || value === undefined || value === '') {
            return Number.NaN;
        }
        var s = String(value).replace(/[^0-9.-]/g, '');
        var n = parseFloat(s);
        return Number.isNaN(n) ? Number.NaN : n;
    }
    // Compare two numbers; NaN sorts to the end for both asc and desc
    function compareNumeric(aNum, bNum, sortOrder) {
        var aNaN = Number.isNaN(aNum);
        var bNaN = Number.isNaN(bNum);
        if (aNaN && bNaN)
            return 0;
        if (aNaN)
            return 1;
        if (bNaN)
            return -1;
        if (aNum < bNum)
            return sortOrder === 'asc' ? -1 : 1;
        if (aNum > bNum)
            return sortOrder === 'asc' ? 1 : -1;
        return 0;
    }
    function getColumnDef(sortField) {
        var columns = config.columns;
        if (!columns)
            return undefined;
        var key = typeof sortField === 'number'
            ? Object.keys(columns)[sortField]
            : sortField;
        return key !== undefined ? columns[key] : undefined;
    }
    function sortData(data, sortField, sortOrder) {
        var columnDef = getColumnDef(sortField);
        var sortValueFn = columnDef === null || columnDef === void 0 ? void 0 : columnDef.sortValue;
        var useNumeric = !sortValueFn && (columnDef === null || columnDef === void 0 ? void 0 : columnDef.sortType) === 'numeric';
        return data.sort(function (a, b) {
            var aRaw = a[sortField];
            var bRaw = b[sortField];
            if (typeof sortValueFn === 'function') {
                var aVal = sortValueFn(aRaw, a);
                var bVal = sortValueFn(bRaw, b);
                var aNum = typeof aVal === 'number' ? aVal : parseNumeric(aVal);
                var bNum = typeof bVal === 'number' ? bVal : parseNumeric(bVal);
                if (typeof aVal === 'number' && typeof bVal === 'number') {
                    return compareNumeric(aNum, bNum, sortOrder);
                }
                return compareValues(aVal, bVal, sortOrder);
            }
            if (useNumeric) {
                var aNum = parseNumeric(aRaw);
                var bNum = parseNumeric(bRaw);
                return compareNumeric(aNum, bNum, sortOrder);
            }
            return compareValues(aRaw, bRaw, sortOrder);
        });
    }
    function toggleSortOrder(currentField, currentOrder, newField) {
        if (currentField === newField) {
            switch (currentOrder) {
                case 'asc':
                    return 'desc';
                case 'desc':
                    return '';
                default:
                    return 'asc';
            }
        }
        return 'asc';
    }
    function setSortIcon(sortField, sortOrder) {
        var _a, _b, _c, _d, _e, _f;
        var baseClass = ((_b = (_a = config.sort) === null || _a === void 0 ? void 0 : _a.classes) === null || _b === void 0 ? void 0 : _b.base) || '';
        var sortClass = sortOrder
            ? sortOrder === 'asc'
                ? ((_d = (_c = config.sort) === null || _c === void 0 ? void 0 : _c.classes) === null || _d === void 0 ? void 0 : _d.asc) || ''
                : ((_f = (_e = config.sort) === null || _e === void 0 ? void 0 : _e.classes) === null || _f === void 0 ? void 0 : _f.desc) || ''
            : '';
        // Clear all headers: remove sort state so only the active column shows highlighted arrow
        var allTh = theadElement.querySelectorAll('th');
        allTh.forEach(function (header) {
            var el = header;
            el.setAttribute('aria-sort', 'none');
            var sortElement = header.querySelector(".".concat(baseClass));
            if (sortElement) {
                sortElement.className = baseClass;
            }
        });
        // Apply sort state to the active column so table.css [aria-sort='asc'] / [aria-sort='desc'] can highlight the arrow
        var th = typeof sortField === 'number'
            ? allTh[sortField]
            : theadElement.querySelector("th[data-kt-datatable-column=\"".concat(String(sortField), "\"], th[data-kt-datatable-column-sort=\"").concat(String(sortField), "\"]"));
        if (th) {
            var sortElement = th.querySelector(".".concat(baseClass));
            if (sortElement) {
                sortElement.className = "".concat(baseClass, " ").concat(sortClass).trim();
            }
            if (sortOrder) {
                th.setAttribute('aria-sort', sortOrder);
            }
            else {
                th.setAttribute('aria-sort', 'none');
            }
        }
    }
    function initSort() {
        if (!theadElement)
            return;
        // Set the initial sort icon
        setSortIcon(getState().sortField, getState().sortOrder);
        // Get all the table headers
        var headers = Array.from(theadElement.querySelectorAll('th'));
        headers.forEach(function (header) {
            var _a, _b;
            // If the sort class is not found, it's not a sortable column
            if (!header.querySelector(".".concat((_b = (_a = config.sort) === null || _a === void 0 ? void 0 : _a.classes) === null || _b === void 0 ? void 0 : _b.base)))
                return;
            // Check if sorting is disabled for this column
            var sortDisabled = header.getAttribute('data-kt-datatable-column-sort') === 'false';
            if (sortDisabled)
                return;
            var sortAttribute = header.getAttribute('data-kt-datatable-column-sort') ||
                header.getAttribute('data-kt-datatable-column');
            var sortField = sortAttribute
                ? sortAttribute
                : header.cellIndex;
            header.addEventListener('click', function () {
                var state = getState();
                var sortOrder = toggleSortOrder(state.sortField, state.sortOrder, sortField);
                setSortIcon(sortField, sortOrder);
                setState(sortField, sortOrder);
                fireEvent('sort', { field: sortField, order: sortOrder });
                dispatchEvent('sort', { field: sortField, order: sortOrder });
                updateData();
            });
        });
    }
    return { initSort: initSort, sortData: sortData, toggleSortOrder: toggleSortOrder, setSortIcon: setSortIcon };
}
