// @keenthemes/ktui/lib/esm/components/clipboard/clipboard.js
/**
 * KTUI - Free & Open-Source Tailwind UI Components by Keenthemes
 * Copyright 2025 by Keenthemes Inc
 */
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g = Object.create((typeof Iterator === "function" ? Iterator : Object).prototype);
    return g.next = verb(0), g["throw"] = verb(1), g["return"] = verb(2), typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
import KTComponent from '../component';
import KTData from '../../helpers/data';
import KTDom from '../../helpers/dom';
var KTClipboard = /** @class */ (function (_super) {
    __extends(KTClipboard, _super);
    function KTClipboard(element, config) {
        if (config === void 0) { config = null; }
        var _this = _super.call(this) || this;
        _this._name = 'clipboard';
        _this._defaultConfig = {
            target: '',
            text: '',
            action: 'copy',
            copiedClass: '',
            successEvent: 'kt.clipboard.success',
            errorEvent: 'kt.clipboard.error',
        };
        _this._config = _this
            ._defaultConfig;
        _this._activateHandler = null;
        // Ensure we don't double bind handlers on the same trigger.
        if (_this._shouldSkipInit(element)) {
            return _this;
        }
        _this._init(element);
        _this._buildConfig(config);
        if (!_this._element)
            return _this;
        _this._activateHandler = _this._handleActivate.bind(_this);
        _this._element.addEventListener('click', _this._activateHandler);
        return _this;
    }
    KTClipboard.prototype._getSuccessEventName = function () {
        var eventName = this._getOption('successEvent');
        return typeof eventName === 'string' && eventName.length > 0
            ? eventName
            : 'kt.clipboard.success';
    };
    KTClipboard.prototype._getErrorEventName = function () {
        var eventName = this._getOption('errorEvent');
        return typeof eventName === 'string' && eventName.length > 0
            ? eventName
            : 'kt.clipboard.error';
    };
    KTClipboard.prototype._getAction = function () {
        var action = this._getOption('action');
        return action === 'cut' ? 'cut' : 'copy';
    };
    KTClipboard.prototype._getTrimmedString = function (optionValue) {
        return typeof optionValue === 'string' ? optionValue.trim() : '';
    };
    KTClipboard.prototype._getPredefinedText = function () {
        var text = this._getOption('text');
        return this._getTrimmedString(text);
    };
    KTClipboard.prototype._getTargetSelector = function () {
        var target = this._getOption('target');
        return this._getTrimmedString(target);
    };
    KTClipboard.prototype._getCopiedClass = function () {
        var copiedClass = this._getOption('copiedClass');
        return this._getTrimmedString(copiedClass);
    };
    KTClipboard.prototype._setCopiedClass = function (shouldSet) {
        var copiedClass = this._getCopiedClass();
        if (!copiedClass || !this._element)
            return;
        // Keep deterministic behavior: remove any previous state before toggling.
        KTDom.removeClass(this._element, copiedClass);
        if (shouldSet) {
            KTDom.addClass(this._element, copiedClass);
        }
    };
    KTClipboard.prototype._isInputLike = function (element) {
        return element.tagName === 'INPUT' || element.tagName === 'TEXTAREA';
    };
    KTClipboard.prototype._readTargetValue = function (target) {
        var _a, _b;
        if (this._isInputLike(target)) {
            return (_a = target.value) !== null && _a !== void 0 ? _a : '';
        }
        return (_b = target.textContent) !== null && _b !== void 0 ? _b : '';
    };
    KTClipboard.prototype._execCommandCopy = function (text) {
        var textarea = document.createElement('textarea');
        textarea.value = text;
        // Avoid scrolling to bottom on iOS/Safari and keep it out of layout.
        textarea.style.position = 'fixed';
        textarea.style.top = '0';
        textarea.style.left = '0';
        textarea.style.opacity = '0';
        textarea.style.pointerEvents = 'none';
        document.body.appendChild(textarea);
        textarea.focus();
        textarea.select();
        // Some browsers require explicit range selection.
        textarea.setSelectionRange(0, textarea.value.length);
        try {
            return document.execCommand('copy');
        }
        catch (_a) {
            return false;
        }
        finally {
            textarea.remove();
        }
    };
    KTClipboard.prototype._writeText = function (text) {
        return __awaiter(this, void 0, void 0, function () {
            var clipboard, writeText, ok;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        clipboard = typeof navigator !== 'undefined' ? navigator.clipboard : null;
                        writeText = clipboard && typeof clipboard.writeText === 'function'
                            ? clipboard.writeText.bind(clipboard)
                            : null;
                        if (!writeText) return [3 /*break*/, 2];
                        return [4 /*yield*/, writeText(text)];
                    case 1:
                        _a.sent();
                        return [2 /*return*/, true];
                    case 2:
                        ok = this._execCommandCopy(text);
                        if (!ok) {
                            throw new Error('Clipboard copy failed.');
                        }
                        return [2 /*return*/, true];
                }
            });
        });
    };
    KTClipboard.prototype._handleActivate = function (event) {
        return __awaiter(this, void 0, void 0, function () {
            var action, successEventName, errorEventName, textFromConfig, targetSelector, hasPredefinedText, targetElForCut, payload, payload, payload, payload, error_1, payload, payload, targetEl, payload, value, payload, payload, payload, error_2, payload;
            var _a;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        event.preventDefault();
                        action = this._getAction();
                        successEventName = this._getSuccessEventName();
                        errorEventName = this._getErrorEventName();
                        textFromConfig = this._getPredefinedText();
                        targetSelector = this._getTargetSelector();
                        hasPredefinedText = Boolean((_a = this._element) === null || _a === void 0 ? void 0 : _a.hasAttribute('data-kt-clipboard-text'));
                        if (!hasPredefinedText) return [3 /*break*/, 5];
                        targetElForCut = null;
                        // `cut` requires an editable target (input/textarea) even when predefined text is used.
                        if (action === 'cut') {
                            if (!targetSelector) {
                                this._setCopiedClass(false);
                                payload = {
                                    action: action,
                                    text: null,
                                    target: null,
                                    error: 'Cut action requires data-kt-clipboard-target pointing to an input/textarea.',
                                };
                                this._fireEvent(errorEventName, payload);
                                this._dispatchEvent(errorEventName, payload);
                                return [2 /*return*/];
                            }
                            targetElForCut = KTDom.getElement(targetSelector);
                            if (!targetElForCut || !this._isInputLike(targetElForCut)) {
                                this._setCopiedClass(false);
                                payload = {
                                    action: action,
                                    text: null,
                                    target: targetSelector,
                                    error: 'Cut action is only supported for input/textarea targets.',
                                };
                                this._fireEvent(errorEventName, payload);
                                this._dispatchEvent(errorEventName, payload);
                                return [2 /*return*/];
                            }
                        }
                        // Treat empty/whitespace-only predefined text as invalid.
                        if (!textFromConfig) {
                            this._setCopiedClass(false);
                            payload = {
                                action: action,
                                text: null,
                                target: targetSelector || null,
                                error: 'Predefined clipboard text is empty.',
                            };
                            this._fireEvent(errorEventName, payload);
                            this._dispatchEvent(errorEventName, payload);
                            return [2 /*return*/];
                        }
                        _b.label = 1;
                    case 1:
                        _b.trys.push([1, 3, , 4]);
                        return [4 /*yield*/, this._writeText(textFromConfig)];
                    case 2:
                        _b.sent();
                        // For `cut`, clear the editable target after successful clipboard write.
                        if (action === 'cut' && targetElForCut) {
                            targetElForCut.value = '';
                            targetElForCut.dispatchEvent(new Event('input', { bubbles: true }));
                        }
                        this._setCopiedClass(true);
                        payload = {
                            action: action,
                            text: textFromConfig,
                            target: targetSelector || null,
                        };
                        this._fireEvent(successEventName, payload);
                        this._dispatchEvent(successEventName, payload);
                        return [3 /*break*/, 4];
                    case 3:
                        error_1 = _b.sent();
                        this._setCopiedClass(false);
                        payload = {
                            action: action,
                            text: textFromConfig,
                            target: targetSelector || null,
                            error: error_1 instanceof Error ? error_1.message : String(error_1 !== null && error_1 !== void 0 ? error_1 : 'Unknown'),
                        };
                        this._fireEvent(errorEventName, payload);
                        this._dispatchEvent(errorEventName, payload);
                        return [3 /*break*/, 4];
                    case 4: return [2 /*return*/];
                    case 5:
                        // No usable predefined text; copy from target.
                        if (!targetSelector) {
                            this._setCopiedClass(false);
                            payload = {
                                action: action,
                                text: null,
                                target: null,
                                error: 'Missing clipboard source (provide data-kt-clipboard-text or data-kt-clipboard-target).',
                            };
                            this._fireEvent(errorEventName, payload);
                            this._dispatchEvent(errorEventName, payload);
                            return [2 /*return*/];
                        }
                        targetEl = KTDom.getElement(targetSelector);
                        if (!targetEl) {
                            this._setCopiedClass(false);
                            payload = {
                                action: action,
                                text: null,
                                target: targetSelector,
                                error: "Clipboard target not found: ".concat(targetSelector),
                            };
                            this._fireEvent(errorEventName, payload);
                            this._dispatchEvent(errorEventName, payload);
                            return [2 /*return*/];
                        }
                        value = this._readTargetValue(targetEl).trim();
                        if (!value) {
                            this._setCopiedClass(false);
                            payload = {
                                action: action,
                                text: null,
                                target: targetSelector,
                                error: 'Target content is empty.',
                            };
                            this._fireEvent(errorEventName, payload);
                            this._dispatchEvent(errorEventName, payload);
                            return [2 /*return*/];
                        }
                        // Cut action: only allowed for input/textarea.
                        if (action === 'cut' && !this._isInputLike(targetEl)) {
                            this._setCopiedClass(false);
                            payload = {
                                action: action,
                                text: null,
                                target: targetSelector,
                                error: 'Cut action is only supported for input/textarea targets.',
                            };
                            this._fireEvent(errorEventName, payload);
                            this._dispatchEvent(errorEventName, payload);
                            return [2 /*return*/];
                        }
                        _b.label = 6;
                    case 6:
                        _b.trys.push([6, 8, , 9]);
                        return [4 /*yield*/, this._writeText(value)];
                    case 7:
                        _b.sent();
                        this._setCopiedClass(true);
                        if (action === 'cut') {
                            targetEl.value = '';
                            targetEl.dispatchEvent(new Event('input', { bubbles: true }));
                        }
                        payload = {
                            action: action,
                            text: value,
                            target: targetSelector,
                        };
                        this._fireEvent(successEventName, payload);
                        this._dispatchEvent(successEventName, payload);
                        return [3 /*break*/, 9];
                    case 8:
                        error_2 = _b.sent();
                        this._setCopiedClass(false);
                        payload = {
                            action: action,
                            text: null,
                            target: targetSelector,
                            error: error_2 instanceof Error ? error_2.message : String(error_2 !== null && error_2 !== void 0 ? error_2 : 'Unknown'),
                        };
                        this._fireEvent(errorEventName, payload);
                        this._dispatchEvent(errorEventName, payload);
                        return [3 /*break*/, 9];
                    case 9: return [2 /*return*/];
                }
            });
        });
    };
    KTClipboard.prototype.dispose = function () {
        if (this._element && this._activateHandler) {
            this._element.removeEventListener('click', this._activateHandler);
        }
        this._activateHandler = null;
        _super.prototype.dispose.call(this);
    };
    KTClipboard.getInstance = function (element) {
        if (!element)
            return null;
        if (KTData.has(element, 'clipboard')) {
            return KTData.get(element, 'clipboard');
        }
        return null;
    };
    KTClipboard.getOrCreateInstance = function (element, config) {
        return (this.getInstance(element) || new KTClipboard(element, config !== null && config !== void 0 ? config : undefined));
    };
    KTClipboard.createInstances = function () {
        document.querySelectorAll('[data-kt-clipboard]').forEach(function (el) {
            new KTClipboard(el);
        });
    };
    KTClipboard.init = function () {
        KTClipboard.createInstances();
    };
    return KTClipboard;
}(KTComponent));
export { KTClipboard };
if (typeof window !== 'undefined') {
    window.KTClipboard = KTClipboard;
}
