// @clerk/react-router/dist/server/keyless/index.js
// src/server/keyless/index.ts
import { createKeylessService } from "@clerk/shared/keyless";
import { clerkClient } from "../clerkClient.js";
import { createFileStorage } from "./fileStorage.js";
var keylessServiceInstance = null;
var keylessInitPromise = null;
function canUseFileSystem() {
  try {
    return typeof process !== "undefined" && typeof process.cwd === "function";
  } catch {
    return false;
  }
}
async function keyless(args, options) {
  if (!canUseFileSystem()) {
    return null;
  }
  if (keylessServiceInstance) {
    return keylessServiceInstance;
  }
  if (keylessInitPromise) {
    return keylessInitPromise;
  }
  keylessInitPromise = (async () => {
    try {
      const storage = await createFileStorage();
      const service = createKeylessService({
        storage,
        api: {
          async createAccountlessApplication(requestHeaders) {
            try {
              return await clerkClient(
                args,
                options
              ).__experimental_accountlessApplications.createAccountlessApplication({
                requestHeaders
              });
            } catch {
              return null;
            }
          },
          async completeOnboarding(requestHeaders) {
            try {
              return await clerkClient(
                args,
                options
              ).__experimental_accountlessApplications.completeAccountlessApplicationOnboarding({
                requestHeaders
              });
            } catch {
              return null;
            }
          }
        },
        framework: "react-router",
        frameworkVersion: "3.0.5"
      });
      keylessServiceInstance = service;
      return service;
    } catch (error) {
      console.warn("[Clerk] Failed to initialize keyless service:", error);
      return null;
    } finally {
      keylessInitPromise = null;
    }
  })();
  return keylessInitPromise;
}
function resetKeylessService() {
  keylessServiceInstance = null;
  keylessInitPromise = null;
}
export {
  keyless,
  resetKeylessService
};
